### PR TITLE
Gripper segment custom models

### PR DESCRIPTION
**Pull request by evelyn-fu**
Add support for using custom SAM2 or Grounding DINO models for gripper segmentation

Failed segmentation using default Grounding DINO + SAM2 models:

Successful segmentation using fine tuned Grounding DINO + SAM2 models:

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -5,3 +5,4 @@ checkpoints
 __pycache__
 outputs
 .vscode
+~/
```
Comment (nepfaff) [inline .gitignore]: Is this needed?

File: README.md [hunk 2]
```diff
@@ -108,6 +108,30 @@ automatic annotations using DINO work well in many cases but can struggle with t
 gripper masks. All downstream object tracking and reconstruction results are sensitive
 to the segmentation quality and thus spending a bit of effort here might be worthwhile.
 
+##### Gripper masking with fine tuned models
+We provide fine tuned networks for SAM2 and GroundingDINO for the segmentation and annotation
+of the gripper used in our provided dataset which can be downloaded from [here](https://mitprod-my.sharepoint.com/personal/nepfaff_mit_edu/_layouts/15/onedrive.aspx?id=%2Fpersonal%2Fnepfaff%5Fmit%5Fedu%2FDocuments%2Fscalable%5Freal2sim%5Fmodel%5Fweights&ga=1).
+
+Please put the downloaded checkpoint files in the `./checkpoints` directory.
```
Comment (nepfaff) [inline README.md:121]: :nit: That directory doesn't currently exist. It might make sense to create it and put a `.gitignore` file inside that includes all files inside this directory apart from the directory itself (you can use `!` to exclude the dir).

File: README.md [hunk 3]
```diff
@@ -108,6 +108,30 @@ automatic annotations using DINO work well in many cases but can struggle with t
...
+When `--txt_prompt` is set to `gripper`, the segmentation script will use the gripper fine tuned
+models for annotation and segmentation.
+
+To fine tune your own object detection model for your gripper, see [these instructions](https://github.com/open-mmlab/mmdetection/blob/main/configs/grounding_dino/README.md)
+from the mmdetection Official Github.
+
+To fine tune your own segmentation model for your gripper, see [these instructions](https://github.com/facebookresearch/sam2/blob/main/training/README.md) for training from the
+SAM2 Official Github.
+
+An example of segmentation failure on the gripper with default models: \
```
Comment (nepfaff) [inline README.md:136]: Could you elaborate a bit here for why it can fail without fine-tuning? It could be worth mentioning that our particular gripper seems to be out of distribution for SAM2, and thus it looses track of it for long videos. This can also be solved with re-prompting it after failure, segmenting the video in parts but fine-tuning removes this extra step.

File: README.md [hunk 4]
```diff
@@ -108,6 +108,30 @@ automatic annotations using DINO work well in many cases but can struggle with t
...
+<img src="assets/mask_sam2_default.png" width="200">
```
Comment (nepfaff) [inline README.md]: :nit: Maybe "_failure" is more appropriate than "_default".

File: README.md [hunk 5]
```diff
@@ -108,6 +108,30 @@ automatic annotations using DINO work well in many cases but can struggle with t
 gripper masks. All downstream object tracking and reconstruction results are sensitive
 to the segmentation quality and thus spending a bit of effort here might be worthwhile.
 
+##### Gripper masking with fine tuned models
```
Comment (nepfaff) [inline README.md:114]: I think it would make sense to also make this an option for the [run_asset_generation.py](https://github.com/nepfaff/scalable-real2sim/blob/main/run_asset_generation.py) pipeline. Maybe add a `store_true` argument to enable it.
In that case, we have to think about where to put the installation instructions. We could keep them here but refer to them [here](https://github.com/nepfaff/scalable-real2sim/blob/main/README.md#segment_moving_obj_datapy) when explaining that argument.

File: README.md [hunk 6]
```diff
@@ -108,6 +108,30 @@ automatic annotations using DINO work well in many cases but can struggle with t
...
+We provide fine tuned networks for SAM2 and GroundingDINO for the segmentation and annotation
+of the gripper used in our provided dataset which can be downloaded from [here](https://mitprod-my.sharepoint.com/personal/nepfaff_mit_edu/_layouts/15/onedrive.aspx?id=%2Fpersonal%2Fnepfaff%5Fmit%5Fedu%2FDocuments%2Fscalable%5Freal2sim%5Fmodel%5Fweights&ga=1).
+
+Please put the downloaded checkpoint files in the `./checkpoints` directory.
+We used mmdetection's implementation to fine tune Grounding DINO. Please see the 
+[mmdetection Official Github](https://github.com/open-mmlab/mmdetection/tree/main)
+for installation instructions.
```
Comment (nepfaff) [inline README.md]: :nit: Could you add a bit more detail such as which `venv` to install that in, so that it works when running our asset generation pipeline.

File: configs/default_runtime.py
Comment (nepfaff) [inline configs/default_runtime.py:1]: :nit: I find it a bit weird to have python files inside a `configs` dir. I would normally expect yaml files or something similar there. Also, these are only for grounding dino fine-tuning. I'm a bit worried that people will lock for general pipeline configs in this folder. Would it be possible to rename this to something more specific (e.g. `finetuned_grounding_dino_utils` or whatever you think an appropriate group name for these file) and move it under `scalable_real2sim/segmentation`?

File: scalable_real2sim/segmentation/segment_moving_object_data.py [hunk 7]
```diff
@@ -14,9 +14,11 @@
 import numpy as np
 import torch
 
+from mmdet.apis import inference_detector, init_detector
```
Comment (nepfaff) [inline scalable_real2sim/segmentation/segment_moving_object_data.py]: This is dangerous as it will now require people to go through the `mm` installation, even if they don't want to use the fine-tuned model. This is problematic as that installation isn't a simple pip install that can be added to the pyproject.toml file.

I would include the new imports into something like this:
```python
try:
    from mmdet.apis import inference_detector, init_detector
    MMDET_AVAILABLE = True
except ImportError:
    logging.warning("... not installed. Finetuned segmentation model not available.")
    MMDET_AVAILABLE = False
```

File: scalable_real2sim/segmentation/segment_moving_object_data.py [hunk 8]
```diff
@@ -112,7 +117,22 @@ def segment_moving_obj_data(
                 "Text prompt must be provided if GUI frames are not specified."
             )
 
-    sam2_checkpoint = "./checkpoints/sam2_hiera_large.pt"
+    if txt_prompt == "gripper":
```
Comment (nepfaff) [inline scalable_real2sim/segmentation/segment_moving_object_data.py:127]: I would again check if the install was successful by adding `MMDET_AVAILABLE` to the if statement. You could print another warning here that mentions that we are falling back to the default model (if `txt_prompt == "gripper" frame).

File: scalable_real2sim/segmentation/segment_moving_object_data.py [hunk 9]
```diff
@@ -112,7 +117,22 @@ def segment_moving_obj_data(
...
+        default_gripper_sam2_path = (
+            "./checkpoints/checkpoint_gripper_finetune_sam2_200epoch_4_1.pt"
+        )
+        if gripper_sam2_path is None:
+            sam2_checkpoint = default_gripper_sam2_path
+        else:
+            sam2_checkpoint = gripper_sam2_path
+
+        if not os.path.exists(sam2_checkpoint):
+            logging.info(
```
Comment (nepfaff) [inline scalable_real2sim/segmentation/segment_moving_object_data.py]: :nit: This should be an error.

File: scalable_real2sim/segmentation/segment_moving_object_data.py [hunk 10]
```diff
@@ -425,28 +468,44 @@ def mouse_callback(event, x, y, flags, param):
...
-            results = processor.post_process_grounded_object_detection(
-                outputs,
-                inputs.input_ids,
-                box_threshold=0.4,
-                text_threshold=0.3,
-                target_sizes=[image.size[::-1]],
-            )
+            if txt_prompt == "gripper":
+                # Use mmdetection api for gripper
+                with autocast(enabled=False):
```
Comment (nepfaff) [inline scalable_real2sim/segmentation/segment_moving_object_data.py]: Do we need `autocost` if we set `enabled=False`?

File: scalable_real2sim/segmentation/segment_moving_object_data.py [hunk 11]
```diff
@@ -653,6 +712,16 @@ def get_dino_boxes(text, frame_idx):
                 except Exception as e:
                     logging.error(f"Error deleting {f}: {e}")
 
+    # save black masks for gripper not found frames
```
Comment (nepfaff) [inline scalable_real2sim/segmentation/segment_moving_object_data.py]: :nit: Starting with lower case breaks the file convention

File: scalable_real2sim/segmentation/segment_moving_object_data.py [hunk 12]
```diff
@@ -425,28 +468,44 @@ def mouse_callback(event, x, y, flags, param):
...
 
-            input_boxes = results[0]["boxes"].cpu().numpy()
-            confidences = results[0]["scores"].cpu().numpy().tolist()
-            class_names = results[0]["labels"]
             return input_boxes, confidences, class_names
 
-        input_boxes, confidences, class_names = get_dino_boxes(
-            txt_prompt, txt_prompt_index
-        )
+        while True:
```
Comment (nepfaff) [inline scalable_real2sim/segmentation/segment_moving_object_data.py]: :nit: Could you add a comment for what this is doing?

File: scalable_real2sim/segmentation/segment_moving_object_data.py (hunk 11 above)
Comment (nepfaff) [reply scalable_real2sim/segmentation/segment_moving_object_data.py]: Also, adding a bit more detail to that comment might be helpful here. Am I understanding it correctly that you are writing black masks until the gripper appears in the frame?

**Review by nepfaff — CHANGES_REQUESTED** [review]

File: .gitignore (hunk 1 above)
Comment (evelyn-fu) [reply .gitignore]: nope, forgot to remove that after adding it in from a local git annoyance

File: scalable_real2sim/segmentation/segment_moving_object_data.py (hunk 10 above)
Comment (evelyn-fu) [reply scalable_real2sim/segmentation/segment_moving_object_data.py]: not sure why this caused an error if I didn't have this here

Comment (nepfaff) [reply scalable_real2sim/segmentation/segment_moving_object_data.py]: In that case, I'd add a comment that mentions that it is needed to avoid errors
